### PR TITLE
Misc refactoring preparing for multi col indices

### PR DESCRIPTION
**Pull request by Centril**
# Description of Changes

Extracted from https://github.com/clockworklabs/SpacetimeDB/pull/439.

- Some code refactoring
- Some field/method renames and additions.

# API and ABI

 - [ ] This is a breaking change to the module ABI
 - [ ] This is a breaking change to the module API
 - [ ] This is a breaking change to the ClientAPI
 - [ ] This is a breaking change to the SDK API

*If the API is breaking, please state below what will break*

# Expected complexity level and risk

2

### CODE REVIEW

File: crates/lib/src/lib.rs [hunk 1]
```diff
@@ -195,16 +196,38 @@ pub struct TypeAlias {
...
+            1 => Ok(IndexType::Hash),
+            _ => Err(()),
+        }
+    }
+}
+
+impl TryFrom<&str> for IndexType {
+    type Error = ();
+    fn try_from(v: &str) -> Result<Self, Self::Error> {
+        match v {
```
Comment (kulakowski) [inline crates/lib/src/lib.rs:223]: I'm concerned about committing to the exact spelling here. Where did BTree vs btree come from?

File: crates/core/src/db/datastore/system_tables.rs [hunk 2]
```diff
@@ -109,10 +110,15 @@ macro_rules! st_fields_enum {
...
+
+        impl From<$ty_name> for NonEmpty<ColId> {
+            fn from(value: $ty_name) -> Self {
+                NonEmpty::new(value.col_id())
+            }
+        }
     }
 }
 
-// WARNING: In order to keep a stable schema, don't change the discriminant of the fields
```
Comment (kulakowski) [inline crates/core/src/db/datastore/system_tables.rs:115]: Why did we lose these comments?

File: crates/core/src/db/datastore/locking_tx_datastore/mod.rs [hunk 3]
```diff
@@ -2486,50 +2566,38 @@ mod tests {
...
-            .sorted_by_key(|x| x.constraint_id)
-            .collect::<Vec<_>>();
-
+        let constraints_rows = query.scan_st_constraints()?;
         #[rustfmt::skip]
         assert_eq!(
             constraints_rows,
             vec![
-                StConstraintRow{ constraint_id: 5.into(), constraint_name: "ct_columns_table_id".to_string(), kind:  ColumnIndexAttribute::INDEXED, table_id: 1.into(), columns: vec![0.into()] },
+                constraint_row(5, "ct_columns_table_id", 1, ColumnIndexAttribute::INDEXED, &[0]),
```
Comment (kulakowski) [inline crates/core/src/db/datastore/locking_tx_datastore/mod.rs:2600]: I mildly don't like that this loses some clarity at call sites, 5 and 1 are like, just numbers now.

File: crates/core/src/db/datastore/locking_tx_datastore/mod.rs [hunk 4]
```diff
@@ -339,6 +334,85 @@ impl SequencesState {
     }
 }
 
+/// Utility to query the system tables and return their concrete table row
+pub struct SystemTableQuery<'a> {
+    db: &'a Inner,
+    ctx: &'a ExecutionContext<'a>,
+}
+
+impl<'a> SystemTableQuery<'a> {
```
Comment (kulakowski) [inline crates/core/src/db/datastore/locking_tx_datastore/mod.rs:343]: Love these.

Comment (cloutiertyler) [inline crates/core/src/db/datastore/locking_tx_datastore/mod.rs:343]: I think these should come with a comment about their performance. Each of these functions requires materializing the full table into a vec.

File: crates/core/src/db/datastore/locking_tx_datastore/mod.rs (hunk 3 above)
Comment (cloutiertyler) [reply crates/core/src/db/datastore/locking_tx_datastore/mod.rs:2600]: I provided the same feedback when @Centril proposes this in another PR. This explicitness is by design. IMO we should revert this change and the other changes like it.

File: crates/lib/src/lib.rs (hunk 1 above)
Comment (mamcx) [reply crates/lib/src/lib.rs:223]: This is for `StIndexRow` where we need to parse the index type.

File: crates/core/src/db/datastore/system_tables.rs (hunk 2 above)
Comment (mamcx) [reply crates/core/src/db/datastore/system_tables.rs:115]: I see them inside the macro `st_fields_enum`

File: crates/core/src/db/datastore/locking_tx_datastore/mod.rs (hunk 4 above)
Comment (mamcx) [reply crates/core/src/db/datastore/locking_tx_datastore/mod.rs:343]: So far this is only for simplify the testing. Probably add a `todo` to change them when we have a better mem-arch?

File: crates/core/src/db/datastore/locking_tx_datastore/mod.rs [hunk 5]
```diff
@@ -2503,165 +2632,99 @@ mod tests {
...
-                column_row(5, 1, "kind", AlgebraicType::U8, false),
-                column_row(5, 2, "epoch", AlgebraicType::U128, false),
-            ]
-        );
-        let index_rows = datastore
-            .iter_mut_tx(&ExecutionContext::default(), &tx, ST_INDEXES_ID)?
-            .map(|x| StIndexRow::try_from(x.view()).unwrap().to_owned())
-            .sorted_by_key(|x| x.index_id)
-            .collect::<Vec<_>>();
+        assert_eq!(query.scan_st_columns()?, map_array([
```
Comment (kulakowski) [inline crates/core/src/db/datastore/locking_tx_datastore/mod.rs:2647]: It seems like there's two kinds of changes to all this test code.

One is fairly mechanical changes due to the renames etc that make up the rest of this PR.

The other is a substantial change to how the datastore exposes system tables to test code.

Is there any other kind of change that happens in this test code?

**Review by mamcx — COMMENTED** [review]
LGTM

**Review by mamcx — APPROVED** [review]
LGTM

**Review by kulakowski — APPROVED** [review]

**Pull request merged**